src/test/java/io/dropwizard/sharding/dao/locktest/LockTest.java

```diff
@@ -254,4 +255,288 @@ public void testPersist_alreadyExistingSame() throws Exception {
+        final String newChildValue = "Newly created child";
+        final String newParentValue = "New parent Value";
+        final DetachedCriteria criteria2 = DetachedCriteria.forClass(SomeOtherObject.class)
+                .add(Restrictions.eq("value", newChildValue));
+        lookupDao.lockAndGetExecutor(p1.getMyId())
+                .selectAndUpdateOrSave(relationDao, criteria2, child -> {
+                    Assert.assertEquals(null, child);
+                    return SomeOtherObject.builder()
+                            .my_id(p1.getMyId())
+                            .value(newChildValue)
+                            .build();
+                })
+                .mutate(parent -> parent.setName(newParentValue))
+                .execute();
+
+        final SomeOtherObject c2 = relationDao.select(p1.getMyId(), criteria2, 0, 1).stream().findFirst().get();
+        Assert.assertEquals(newChildValue, c2.getValue());
+        Assert.assertNotEquals(c1.getId(), c2.getId());
+        Assert.assertEquals(newParentValue,lookupDao.get(p1Id).get().getName());
+        Assert.assertEquals(p1Id,lookupDao.get(p1Id).get().getMyId());
+    }
+
+    @Test
+    public void testScrollAndUpdate() throws Exception {
+        final String p1Id = "0";
+        final SomeLookupObject p1 = SomeLookupObject.builder()
+                .myId(p1Id)
+                .name("Parent 1")
+                .build();
+
+        final SomeOtherObject c1 = relationDao.save(p1.getMyId(), SomeOtherObject.builder()
+                .my_id(p1.getMyId())
+                .value("Hello1")
+                .build()).get();
+
+        final SomeOtherObject c2 = relationDao.save(p1.getMyId(), SomeOtherObject.builder()
+                .my_id(p1.getMyId())
+                .value("Hello2")
+                .build()).get();
+
+        final String p2Id = "1";
+        final SomeLookupObject p2 = SomeLookupObject.builder()
+                .myId(p2Id)
+                .name("Parent 2")
+                .build();
+
+        final SomeOtherObject c3 = relationDao.save(p2.getMyId(), SomeOtherObject.builder()
+                .my_id(p2.getMyId())
+                .value("Hello3")
+                .build()).get();
+
+        lookupDao.save(p1);
+        lookupDao.save(p2);
+
+        //test full update
+        final DetachedCriteria criteria = DetachedCriteria.forClass(SomeOtherObject.class)
+                .add(Restrictions.eq("my_id", p1.getMyId()));
+
+        final String childModifiedValue = "Hello Modified";
+        final String parentModifiedValue = "Parent Changed";
+
+        lookupDao.lockAndGetExecutor(p1.getMyId())
+                .scrollAndUpdate(relationDao, criteria,ScrollMode.FORWARD_ONLY, scrollableResults -> {
+                    final List<SomeOtherObject> modified = new ArrayList<>();
+                    while(scrollableResults.next()) {
+                        final SomeOtherObject entity = (SomeOtherObject) scrollableResults.get(0);
```

Comment:
It's bad if client has to typecast. Client code should only get strongly typed object.

Reply:
Because the code can be written only once, it is sufficient to create a function like:
`update(dao, criteria, Function<T,T> updater)` along side the existing `update(RelationalDao<U> relationalDao, Object id, Function<U, U> handler)`. Internally the function will scroll through all the results and call the updater.